File: clipsync-win/ClipSync/Net/PeerRegistry.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace ClipSync.Net;

public sealed record Peer(string DidHex, string Name, PeerState State);
/// Looking: a trusted peer we have seen advertised but not yet connected
/// to, and have no evidence is unreachable. It is the honest label between
/// discovery and the first Hello — calling that window "Offline" reported a
/// guess as a fact for however long the connect took.
public enum PeerState { Online, Pending, Looking, Offline }

public sealed class PeerRegistry
{
    public Action<ClipboardItem>? OnRemoteItem;
    public Action<IReadOnlyList<Peer>>? OnChange;

    private readonly ConcurrentDictionary<string, PeerConnection> _connections = new();
    private readonly ConcurrentDictionary<string, Peer> _discovered = new();
    /// When each trusted peer was first seen this session; Looking expires
    /// LookingWindow after that if nothing has connected.
    private readonly ConcurrentDictionary<string, DateTime> _firstSeen = new();
    /// Peers whose every advertised address refused a connect. Cleared when
    /// a connection does come up.
    private readonly ConcurrentDictionary<string, byte> _unreachable = new();
    private readonly string _localDidHex;

    private static readonly TimeSpan LookingWindow = TimeSpan.FromSeconds(30);

    public PeerRegistry(string localDidHex) { _localDidHex = localDidHex; }

    public bool IsConnected(string didHex) => _connections.ContainsKey(didHex.ToLowerInvariant());

    /// Called by Discovery whenever a service instance is seen on the network.
    public void OnDiscovered(string didHex, string name, bool trusted)
    {
        var key = didHex.ToLowerInvariant();
        var state = trusted ? PeerState.Looking : PeerState.Pending;
        _discovered[key] = new Peer(key, name, state);
        if (trusted && _firstSeen.TryAdd(key, DateTime.UtcNow))
        {
            // Nothing else re-emits when the window lapses, so schedule the
            // flip to Offline ourselves; harmless if it connected meanwhile.
            _ = Task.Delay(LookingWindow).ContinueWith(_ => Emit());
        }
        Security.Identity.Log($"PeerRegistry.OnDiscovered: {name} did={key} trusted={trusted} state={state}");
        Emit();
    }

    /// Called by Discovery when a connect attempt has tried every address
    /// the peer advertised and failed: that is real evidence, so stop
    /// saying Looking.
    public void MarkUnreachable(string didHex)
    {
        var key = didHex.ToLowerInvariant();
        if (_connections.ContainsKey(key)) return;
        if (_unreachable.TryAdd(key, 0))
        {
            Security.Identity.Log($"PeerRegistry.MarkUnreachable: {key[..8]}");
            Emit();
        }
    }

    /// The state to display for a discovered, not-connected peer.
    private PeerState ResolveState(string key, Peer discovered)
    {
        if (discovered.State != PeerState.Looking) return discovered.State;
        if (_unreachable.ContainsKey(key)) return PeerState.Offline;
        if (_firstSeen.TryGetValue(key, out var seen) && DateTime.UtcNow - seen < LookingWindow)
            return PeerState.Looking;
        return PeerState.Offline;
    }

    public void Adopt(PeerConnection pc)
    {
        pc.OnItem = item =>
        {
            try { OnRemoteItem?.Invoke(item); }
            catch (Exception ex) { Security.Identity.Log($"OnRemoteItem error: {ex.Message}"); }
        };
        pc.OnReady = () =>
        {
            if (pc.PeerDid is not { } did) return;
            var hex = Convert.ToHexString(did).ToLowerInvariant();
            _unreachable.TryRemove(hex, out _);
            while (true)
            {
                if (_connections.TryAdd(hex, pc)) break;
                if (!_connections.TryGetValue(hex, out var existing)) continue;
                if (ReferenceEquals(existing, pc)) break;
                // Simultaneous connect from both sides. Both ends must keep
                // the *same* connection, so tie-break deterministically:
                // keep the one where the lower-DID device is the TLS client.
                var keepClient = string.CompareOrdinal(_localDidHex, hex) < 0;
                if ((pc.Role == PeerRole.Client) == keepClient)
                {
                    if (_connections.TryUpdate(hex, pc, existing))
                    {
                        Security.Identity.Log($"PeerRegistry: replacing duplicate connection for {hex}");
                        existing.Close();
                        break;
                    }
                }
                else
                {
                    Security.Identity.Log($"PeerRegistry: dropping duplicate connection for {hex}");
                    pc.Close();
                    return;
                }
            }
            Emit();
        };
        pc.OnClose = () =>
        {
            if (pc.PeerDid is { } did)
            {
                var hex = Convert.ToHexString(did).ToLowerInvariant();
                // Only remove if this pc is still the registered connection —
                // a replaced duplicate must not evict its successor.
                _connections.TryRemove(new KeyValuePair<string, PeerConnection>(hex, pc));
            }
            Emit();
        };
    }

    public IReadOnlyList<Peer> GetAll()
    {
        var list = new List<Peer>();
        var connected = new HashSet<string>();
        Security.Identity.Log($"GetAll: _connections.Count={_connections.Count}");
        foreach (var (hex, pc) in _connections)
        {
            Security.Identity.Log($"  connected: {hex} name={pc.PeerName}");
            list.Add(new Peer(hex, pc.PeerName ?? "Peer", PeerState.Online));
            connected.Add(hex);
        }
        foreach (var (hex, peer) in _discovered)
        {
            if (!connected.Contains(hex))
                list.Add(peer with { State = ResolveState(hex, peer) });
        }
        Security.Identity.Log($"GetAll: returning {list.Count} peers");
        return list;
    }

    /// Consulted for each peer before sending. Null means send to everyone, which
    /// is what this class did before pausing existed and what it still does
    /// if nobody wires it up. Keeping the decision outside means the registry
    /// does not need to know what a pause is, and the global and per-peer
    /// cases both arrive through one predicate.
    public Func<string, bool>? ShouldSendTo;

    public void Broadcast(ClipboardItem item)
    {
        foreach (var (hex, pc) in _connections)
        {
            // Both branches log. A skip needs saying, or a user watches
            // nothing arrive with no way to tell a pause from a broken link;
            // and a send needs saying too, or the absence of a skip line is
            // indistinguishable from the item never reaching this method.
            if (ShouldSendTo is { } gate && !gate(hex))
            {
                Security.Identity.Log($"Broadcast: not sending to {hex[..8]} (paused)");
                continue;
            }
            Security.Identity.Log($"Broadcast: sending to {hex[..8]}");
            _ = pc.SendItemAsync(item);
        }
    }

    private void Emit()
    {
        var list = new List<Peer>();
        var connected = new HashSet<string>();
        foreach (var (hex, pc) in _connections)
        {
            list.Add(new Peer(hex, pc.PeerName ?? "Peer", PeerState.Online));
            connected.Add(hex);
        }
        foreach (var (hex, peer) in _discovered)
        {
            if (!connected.Contains(hex))
                list.Add(peer with { State = ResolveState(hex, peer) });
        }
        Security.Identity.Log($"PeerRegistry.Emit: {list.Count} peers, OnChange={OnChange is not null}");
        OnChange?.Invoke(list);
    }
}

File: clipsync-mac/Sources/ClipSync/Net/PeerRegistry.swift
import Foundation
import Network

struct Peer: Identifiable, Equatable {
    var id: String { didHex }
    let didHex: String
    let name: String
    let state: State
    /// App version a connected peer reported in its Hello; nil for peers
    /// that aren't connected or predate the field.
    var version: String? = nil
    /// First 8 hex chars of the SPKI fingerprint — shown in the TOFU
    /// pair UI so the user can eyeball-verify it matches on both sides.
    var fingerprintShort: String { String(didHex.prefix(8)) }
    /// looking: a trusted peer seen advertised but not yet connected, with
    /// no evidence it is unreachable — the honest label between discovery
    /// and the first Hello. Calling that window "offline" would report a
    /// guess as a fact for however long the connect takes.
    enum State: Equatable { case online, pending, looking, offline }

    func with(state: State) -> Peer {
        Peer(didHex: didHex, name: name, state: state, version: version)
    }
}

final class PeerRegistry {
    var onRemoteItem: ((ClipboardItem) -> Void)?
    var onChange: (([Peer]) -> Void)?
    /// Fired when the user clicks Trust on a pending peer. Discovery
    /// wires this up to persist to the trust store and kick off a
    /// real mTLS connect using the saved endpoint.
    var onTrustRequested: ((String, String, NWEndpoint) -> Void)?
    /// Our own DID hex, used to break ties when both sides connect at once.
    var localDidHex: String = ""

    private var connections: [String: PeerConnection] = [:]       // didHex → live mTLS conn
    private var pending: [String: (peer: Peer, endpoint: NWEndpoint)] = [:]
    /// Trusted peers seen advertised but not yet connected, shown as
    /// "Looking…". Keyed by didHex.
    private var looking: [String: Peer] = [:]
    /// When each was first seen this session; "Looking…" lapses to
    /// "Offline" LookingWindow after that if nothing connects.
    private var firstSeen: [String: Date] = [:]
    /// Peers whose connect attempt reached every advertised address and
    /// failed — evidence they are off. Cleared when a connection comes up.
    private var unreachable: Set<String> = []

    private static let lookingWindow: TimeInterval = 30

    // All state above is touched only on the main queue (Discovery runs
    // there and every connection callback is dispatched there), so no lock.

    func adopt(_ pc: PeerConnection) {
        pc.onItem = { [weak self] in self?.onRemoteItem?($0) }
        pc.onReady = { [weak self] in self?.registerReady(pc) }
        pc.onClose = { [weak self] in self?.unregister(pc) }
    }

    private func registerReady(_ pc: PeerConnection) {
        guard let did = pc.peerDid else { return }
        let hex = did.map { String(format: "%02x", $0) }.joined()
        if let existing = connections[hex], existing !== pc {
            // Simultaneous connect from both sides. Both ends must keep the
            // *same* connection, so tie-break deterministically: keep the
            // one where the lower-DID device is the TLS client.
            let keepClient = localDidHex < hex
            if (pc.role == .client) == keepClient {
                connections[hex] = pc
                existing.close()
            } else {
                pc.close()
                return
            }
        } else {
            connections[hex] = pc
        }
        pending.removeValue(forKey: hex)
        // A live link is the strongest evidence of reachability.
        unreachable.remove(hex)
        emit()
    }

    private func unregister(_ pc: PeerConnection) {
        if let did = pc.peerDid {
            let hex = did.map { String(format: "%02x", $0) }.joined()
            // Only remove if this pc is still the registered connection —
            // a replaced duplicate must not evict its successor.
            if connections[hex] === pc {
                connections.removeValue(forKey: hex)
            }
        }
        emit()
    }

    func isConnected(didHex: String) -> Bool { connections[didHex] != nil }

    func notePending(name: String, didHex: String?, endpoint: NWEndpoint) {
        guard let didHex, connections[didHex] == nil else { return }
        pending[didHex] = (
            Peer(didHex: didHex, name: name, state: .pending),
            endpoint
        )
        emit()
    }

    /// A trusted peer has been discovered but is not yet connected. Show it
    /// as "Looking…" rather than omitting it, so a peer that exists but
    /// hasn't linked up is visible. Called each time it is seen advertised.
    func noteLooking(name: String, didHex: String) {
        let key = didHex.lowercased()
        guard connections[key] == nil else { return }
        looking[key] = Peer(didHex: key, name: name, state: .looking)
        if firstSeen[key] == nil {
            firstSeen[key] = Date()
            // Nothing else re-emits when the window lapses, so schedule the
            // flip to Offline; harmless if a connection arrives meanwhile.
            DispatchQueue.main.asyncAfter(deadline: .now() + Self.lookingWindow) { [weak self] in
                self?.emit()
            }
        }
        emit()
    }

    /// A connect attempt tried every advertised address and never reached
    /// the peer: real evidence it is off, so stop saying "Looking…".
    func markUnreachable(didHex: String) {
        let key = didHex.lowercased()
        guard connections[key] == nil else { return }
        if unreachable.insert(key).inserted { emit() }
    }

    /// State to display for a discovered, not-connected trusted peer:
    /// "Looking…" until either a connect fails outright or the window
    /// lapses, then "Offline". After that first window the state is known,
    /// so a peer that drops later reads "Offline" at once.
    private func resolveState(_ key: String, _ peer: Peer) -> Peer.State {
        guard peer.state == .looking else { return peer.state }
        if unreachable.contains(key) { return .offline }
        if let seen = firstSeen[key], Date().timeIntervalSince(seen) < Self.lookingWindow {
            return .looking
        }
        return .offline
    }

    /// Called from the menu-bar "Trust" button.
    func trust(didHex: String) {
        guard let entry = pending[didHex] else { return }
        onTrustRequested?(didHex, entry.peer.name, entry.endpoint)
    }

    /// Consulted for each peer before sending. Nil means send to
    /// everyone, which is what this class did before pausing existed and
    /// what it still does if nobody wires it up. Keeping the decision
    /// outside means the registry never learns what a pause is, and the
    /// global and per-peer cases both arrive through one predicate.
    var shouldSendTo: ((String) -> Bool)?

    func broadcast(_ item: ClipboardItem) {
        for (hex, pc) in connections {
            // Both branches log. A skip needs saying, or a user watches
            // nothing arrive with no way to tell a pause from a broken
            // link; and a send needs saying too, or the absence of a skip
            // line is indistinguishable from the item never reaching this
            // method.
            if let gate = shouldSendTo, !gate(hex) {
                NSLog("Broadcast: not sending to %@ (paused)", String(hex.prefix(8)))
                continue
            }
            NSLog("Broadcast: sending to %@", String(hex.prefix(8)))
            pc.send(item: item)
        }
    }

    private func emit() {
        var list: [Peer] = []
        var shown = Set<String>()
        for (hex, pc) in connections {
            list.append(Peer(didHex: hex, name: pc.peerName ?? "Peer", state: .online,
                             version: pc.peerVersion))
            shown.insert(hex)
        }
        for (hex, entry) in pending where !shown.contains(hex) {
            list.append(entry.peer)
            shown.insert(hex)
        }
        for (hex, peer) in looking where !shown.contains(hex) {
            list.append(peer.with(state: resolveState(hex, peer)))
            shown.insert(hex)
        }
        onChange?(list)
    }
}
